graph-builder/src/graph.rs

```diff
@@ -77,15 +79,29 @@ pub struct State {
     json: Arc<RwLock<String>>,
     /// Query parameters that must be present in all client requests.
     mandatory_params: HashSet<String>,
+    live: Arc<RwLock<bool>>,
 }
 
 impl State {
-    pub fn new(json: Arc<RwLock<String>>, mandatory_params: HashSet<String>) -> State {
+    pub fn new(
+        json: Arc<RwLock<String>>,
+        mandatory_params: HashSet<String>,
+        live: Arc<RwLock<bool>>,
+    ) -> State {
         State {
             json,
             mandatory_params,
+            live,
         }
     }
+
+    pub fn is_live(&self) -> bool {
+        *self.live.read()
+    }
+
+    pub fn is_ready(&self) -> bool {
+        (UPSTREAM_SCRAPES.get() - UPSTREAM_ERRORS.get()) > 0.0
```

Comment:
This logic looks a bit strange to me.
`UPSTREAM_SCRAPES` can normally go up quite a bit (it's ~15k on current prod), so it will take a huge amount of runs for `UPSTREAM_ERRORS` to balance that and signal not-ready (if it ever manages to). Similarly, after a long strike of errors it will take a proportional strike of non-errors to reach ready again.

Reply:
in case of an error both will increase. What this does is simply see if there was *ever* one successful scrape.

Reply:
Ah, that's clever, I misread the logic!
Let's keep this then, just add a docstring or a comment to make that logic explicit.